pkg/infra/state_reader/state_reader.go

```diff
@@ -1,62 +1,331 @@
+	return errs
+}
+
+func (p stateReader) loadProperties(state stateconverter.State, pc propertyCorrelation) error {
+	var errs error
+	for id, properties := range state {
+		resource, err := p.graph.Vertex(id)
+		if err != nil {
+			errs = errors.Join(errs, err)
+			continue
+		}
+		for key, value := range properties {
+			errs = errors.Join(errs, pc.setProperty(resource, key, value))
+		}
+	}
+	return errs
+}
+
+func (p propertyCorrelator) setProperty(
+	resource *construct.Resource,
+	property string,
+	value any,
+) error {
+	edges := make([]*construct.Edge, 0)
+	rt, err := p.ctx.KnowledgeBase.GetResourceTemplate(resource.ID)
+	if err != nil {
+		return err
+	} else if rt == nil {
+		return fmt.Errorf("resource template not found for resource %s", resource.ID)
+	}
+	parts := strings.Split(property, "#")
+	property = parts[0]
+	prop := rt.GetProperty(property)
+	if prop == nil {
+		return fmt.Errorf("property %s not found in resource template %s", property, resource.ID)
+	}
+	opRule := prop.Details().OperationalRule
+	if opRule == nil || len(opRule.Step.Resources) == 0 {
+		return resource.SetProperty(property, value)
+	}
+	var ref string
+	if len(parts) > 1 {
+		ref = parts[1]
+	} else {
+		ref = opRule.Step.UsePropertyRef
+	}
+
+	switch rval := reflect.ValueOf(value); rval.Kind() {
+	case reflect.String:
+		edge, pref, err := p.checkValue(opRule.Step, value.(string), resource.ID, ref)
+		if err != nil {
+			return err
+		}
+		if edge != nil {
+			edges = append(edges, edge)
+		}
+		if pref != nil {
+			switch prop.(type) {
+			case *properties.ResourceProperty:
+				err = prop.SetProperty(resource, pref.Resource)
+				if err != nil {
+					return err
+				}
+			default:
+				err = prop.SetProperty(resource, pref)
+				if err != nil {
+					return err
+				}
+			}
+
+		}
+	case reflect.Slice, reflect.Array:
+		var val []any
+		for i := 0; i < rval.Len(); i++ {
+			edge, pref, err := p.checkValue(opRule.Step, rval.Index(i).Interface().(string), resource.ID, ref)
+			if err != nil {
+				return err
+			}
+			if edge != nil {
+				edges = append(edges, edge)
+			}
+			if pref != nil {
+				val = append(val, *pref)
+			}
+		}
+		collectionProp, ok := prop.(knowledgebase.CollectionProperty)
+		if !ok {
+			return fmt.Errorf("property %s is not a collection property", property)
+		}
+		switch collectionProp.Item().(type) {
+		case *properties.ResourceProperty:
+			resources := make([]construct.ResourceId, 0)
+			for _, v := range val {
+				resources = append(resources, v.(construct.PropertyRef).Resource)
+			}
+			err = prop.SetProperty(resource, resources)
+			if err != nil {
+				return err
+			}
+		default:
+			err = prop.SetProperty(resource, val)
 			if err != nil {
-				return nil, err
+				return err
 			}
 		}
-		err = returnGraph.AddVertex(resource)
+	}
+	for _, edge := range edges {
+		err := p.ctx.Graph.AddEdge(edge.Source, edge.Target)
+		if err != nil {
+			return err
+		}
+	}
+	return nil
+}
+
+func (p propertyCorrelator) checkValue(
```

Comment:
nit: this function could use either a better name or some documentation on what it does. Seems like it does more than checking, since I see a `SetProperty` and `AddVertex` (write operations).